### PR TITLE
Task 70711 graceful shutdown

### PR DESCRIPTION
**Pull request by mszymeshop**
### What
Ensure all messages have been processed before the source reader (for the message) is deleted.

### How

1. Director service informs Reader service of the intent to gracefully shut down
2. Reader service stops retrieving new messages and awaits the Event Handler to inform on message processing completion
3. Reader service identifies the final message has been processed and initiates self-delete

### Testing method
When testing I have been constantly sending messages in the background and changing the config of a webhook. I would set it to target endpoint A and use API to reload config. Then after a few seconds, I would change the setting to target endpoint B and use API to reload config.

### Problems
Quite often the sum of processed and delivered messages is greater than the sum of generated messages and pushed to Service Bus. They are never lower, which is good. However, some messages can be delivered twice and at the moment I cannot pinpoint it to the area in the code. If you could take a look, maybe you can spot something.

### CODE REVIEW

File: src/CaptainHook.EventReaderService/EventReaderService.cs [hunk 1]
```diff
@@ -199,7 +217,7 @@ protected override async Task RunAsync(CancellationToken cancellationToken)
                 _freeHandlers = Enumerable.Range(1, HandlerCount).ToConcurrentQueue();
                 await SetupServiceBus();
 
-                while (!cancellationToken.IsCancellationRequested)
+                while (!cancellationToken.IsCancellationRequested && _isShuttingDown == 0)
```
[the remark this answers is not in the document]
Comment (mszymeshop) [reply src/CaptainHook.EventReaderService/EventReaderService.cs:220]: @dguerin I have actually started with this approach to use the cancellation token and unfortunately, this interferes with remoting listeners which are closed in parallel (without waiting for RunAsync to finish): https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-lifecycle#stateful-service-shutdown. As a result I wasn't receiving the CompleteMessage communication from Handlers.

File: src/CaptainHook.EventReaderService/EventReaderService.cs [hunk 2]
```diff
@@ -403,8 +430,31 @@ public async Task CompleteMessageAsync(MessageData messageData, bool messageDeli
...
         }
+
+        public Task InitializeGracefulShutdown()
+        {
+            Interlocked.Exchange(ref _isShuttingDown, 1);
+
+            _bigBrother.Publish(new GracefulShutdownRequestedForReaderEvent(Context.ServiceName.AbsoluteUri, _initData.EventType));
+            var shutdownCompletedTimedEvent = new GracefulShutdownCompletedForReaderEvent(Context.ServiceName.AbsoluteUri, _initData.EventType);
+
+            ThreadPool.QueueUserWorkItem(state =>
```
[the remark this answers is not in the document]
Comment (mszymeshop) [reply src/CaptainHook.EventReaderService/EventReaderService.cs:448]: @dguerin for some reason I was unable to communicate back to Director Service from the Reader that the processing has been done and finally decided for this self-delete approach. That notification to DS would need to come from CompleteMessages method which on itself is called through remoting. I am not sure, but maybe that was the problem? It was a weird programmatically approach that I have come across, wasn't able to resolve it.

File: src/CaptainHook.DirectorService/Infrastructure/ReaderServicesManager.cs [hunk 3]
```diff
@@ -100,13 +102,21 @@ private async Task CreateReaderServicesAsync(IDictionary<string, SubscriberConfi
...
-            foreach (var oldName in oldNames)
+            IEnumerable<Task> CreateShutdownTasksForOldNames()
             {
-                if (cancellationToken.IsCancellationRequested) return;
+                foreach (var oldName in oldNames)
+                {
+                    if (cancellationToken.IsCancellationRequested) yield break;
+
+                    var serviceToShutdownUri = new Uri(oldName);
+                    var task = ServiceProxy.Create<IEventReaderService>(serviceToShutdownUri).InitializeGracefulShutdown();
```
[the remark this answers is not in the document]
Comment (chomick) [reply src/CaptainHook.DirectorService/Infrastructure/ReaderServicesManager.cs:112]: Agree on abstracting the serviceproxy.

File: src/CaptainHook.EventReaderService/EventReaderService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chomick) [reply src/CaptainHook.EventReaderService/EventReaderService.cs:448]: Any reason we are not using Task.Run () here? They are equivalent but Task is the newer form of doing the job + the exceptions are handled better  in that case, when using await on it.
Also, I wonder if we could make this async method and return the newly created task.

**Pull request closed without merging**